crates/iota-genesis-builder/src/lib.rs

```diff
@@ -253,75 +265,116 @@ impl Builder {
-    /// This merges conditionally the cached token distribution
-    /// (i.e. `self.token_distribution_schedule`)  with the genesis stake
-    /// resulting from the migrated state.
-    ///
-    /// If the cached token distribution schedule contains timelocked stake, it
-    /// is assumed that the genesis stake is already merged and no operation
-    /// is performed. This is the case where we load a [`Builder`] from disk
-    /// that has already built genesis with the migrated state.
+    /// There are 6 cases for evaluating this:
+    /// 1. The genesis is built WITHOUT migration
+    ///    1. and a schedule is given as input -> then just use the input
+    ///       schedule;
+    ///    2. and the schedule is NOT given as input -> then instantiate a
+    ///       default token distribution schedule for a genesis without
+    ///       migration.
+    /// 2. The genesis is built with migration,
+    ///    1. and token distribution schedule is given as input
+    ///       1. if the token distribution schedule contains a timelocked stake
+    ///          -> then just use the input schedule, because it was initialized
+    ///          for migration before this execution (this is the case where we
+    ///          load a [`Builder`] from disk that has already built genesis
+    ///          with the migrated state.);
+    ///       2. if the token distribution schedule does NOT contain any
+    ///          timelocked stake -> then fetch the cached the genesis stake and
+    ///          merge it to the token distribution schedule;
+    ///    2. and token distribution schedule is NOT given as input -> then
+    ///       fetch the cached genesis stake and initialize a new token
+    ///       distribution schedule with it.
     fn resolve_token_distribution_schedule(&mut self) -> TokenDistributionSchedule {
-        let validator_addresses = self.validators.values().map(|v| v.info.iota_address());
-        let token_distribution_schedule = self.token_distribution_schedule.take();
+        let is_genesis_with_migration = !self.migration_objects.is_empty();
         let stardust_total_supply_nanos =
             self.migration_sources.len() as u64 * STARDUST_TOTAL_SUPPLY_NANOS;
-        if self.genesis_stake.is_empty() {
-            token_distribution_schedule.unwrap_or_else(|| {
-                TokenDistributionSchedule::new_for_validators_with_default_allocation(
-                    validator_addresses,
-                )
-            })
-        } else if let Some(schedule) = token_distribution_schedule {
-            if schedule.contains_timelocked_stake() {
-                // Genesis stake is already included
-                schedule
+
+        if let Some(schedule_without_migration) = self.token_distribution_schedule.take() {
+            if !is_genesis_with_migration || schedule_without_migration.contains_timelocked_stake()
+            {
+                // Case 1.1 and 2.1.1
+                schedule_without_migration
```

Comment:
As you suggest also in the documentation case 2.1.1 includes the migration so `schedule_without_migration` is rather misleading a name. So I suggest reverting back to `schedule`.

Reply:
Correct, changed